nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/PutAWSIoT.java

```diff
@@ -0,0 +1,141 @@
+                    "should retain a copy of the message. The message will then be transmitted to new " +
+                    "subscribers to a topic that matches the message topic. For subscribers registering " +
+                    "a new subscription, the flag being true indicates that the received message is not " +
+                    "a new one, but one that has been retained by the MQTT server.")
+            .required(true)
+            .defaultValue(PROP_DEFAULT_RETAINED.toString())
+            .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
+            .build();
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(
+                    PROP_QOS,
+                    PROP_TOPIC,
+                    PROP_RETAINED,
+                    PROP_ENDPOINT,
+                    PROP_KEEPALIVE,
+                    PROP_CLIENT,
+                    AWS_CREDENTIALS_PROVIDER_SERVICE,
+                    TIMEOUT,
+                    REGION));
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return Collections.singleton(REL_SUCCESS);
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        shouldRetain = context.getProperty(PROP_RETAINED).isSet() ? context.getProperty(PROP_RETAINED).asBoolean() : PROP_DEFAULT_RETAINED;
+        init(context);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        // check if MQTT-connection is about to expire
+        if (isConnectionAboutToExpire()) {
+            // renew connection
+            mqttClient = connect(context);
+        }
+        // get flowfile
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+        Map<String, String> attributes = flowFile.getAttributes();
+        // if provided override MQTT configuration with values from the corresponding message attributes
+        String topic = attributes.containsKey(ATTR_NAME_TOPIC) ? attributes.get(ATTR_NAME_TOPIC) : awsTopic;
+        Integer qos = attributes.containsKey(ATTR_NAME_QOS) ? Integer.parseInt(attributes.get(ATTR_NAME_QOS)) : awsQos;
+        Boolean retained = attributes.containsKey(ATTR_NAME_RETAINED) ? Boolean.parseBoolean(attributes.get(ATTR_NAME_RETAINED)) : shouldRetain;
+        // get message content
+        final ByteArrayOutputStream fileContentStream = new ByteArrayOutputStream();
+        session.exportTo(flowFile, fileContentStream);
+
+        try {
+            // publish messages to mqtt-topic(s)
+            mqttClient.publish(topic, fileContentStream.toByteArray(), qos, retained);
+            session.transfer(flowFile, REL_SUCCESS);
```

Comment:
Since the mqttClient is working async, there is the chance that the client will attempt to publish, the processor will then transfer the flowfile and then the publish will fail, leading to data loss (acceptable with QoS 0 but not 1). The best way to handle this is for QoS 1 to wait until the message is delivered, like the default Paho MQTT client does: 

https://github.com/eclipse/paho.mqtt.java/blob/master/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/MqttClient.java#L361